File: chapter_02/new.c


struct class {
	size_t size;
	void * (* ctor) (void * self, va_list *app);
	void * (* dtor) (void * self);
	void * (* clone) (const void * self);
	int (* differ)(const void * self, const void * b);
};

void * new (const void * class_arg, ... ) {
	const struct class * c = class_arg;
	void * p = calloc(1, c-> size);

	assert(p);
	*(const struct class **)p = c;

	if(c -> ctor) {
		va_list ap;

		va_start(ap, class_arg);
		p = c -> ctor(p, &ap);
		va_end(ap);
	}

	return p;
}

File: chapter_02/new.r
#ifndef CLASS_R
#define CLASS_R

struct class {
	/* new needs to know how much mem to allocate */
	size_t size;

	/* new calls this to initialize obj
	 * (may install dynamic mem for itself) */
	void * (* ctor) (void * self, va_list *app);
	/* delete calls this to properly destroy 
	 * the obj, with dynamically allocated mem */
	void * (* dtor) (void * self);
	void * (* clone) (const void * self);
	/* Not a simple case of pointer difference 
	 * anymore. objects could be math style
	 * matrices and have same numbers, but are 
	 * different objects (copies). Like 
	 * python deepcopy, I think */
	int (* differ)(const void * self, const void * b);
};

#endif
